third_party/move/move-model/src/ty.rs

```diff
@@ -282,6 +282,19 @@ impl Type {
         }
     }
 
+    pub fn require_struct_opt(&self) -> Option<(ModuleId, StructId, &[Type])> {
+        if let Type::Struct(mid, sid, targs) = self {
+            Some((*mid, *sid, targs.as_slice()))
+        } else if let Type::TypeParameter(_) = self {
```

Comment:
I think you do not need to have this check for parameters. Just return None (no panic) if its not a struct. You can then also reduce code duplication, and define `require_struct` in terms of `require_struct_opt`.

You do not need this extra check because you already type checking whether its a parameter nor struct.

Reply:
I added the check for parameters to handle the case such as `exists<u64>`.

Reply:
But this is already type checker, so not sure whether you need the panic here.

The least thing to do if you want to keep this function is rename it to `require_struct_or_type_param_opt`. Otherwise it is very confusing because `require_struct` already exists -- but this here is not the same.

Reply:
Yes, you are right. I fixed it and also refactored other functions using `_opt` to reduce code duplication.